File: src/ChatForm.cpp
// Copyright © 2008-2014 Pioneer Developers. See AUTHORS.txt for details
// Licensed under the terms of the GPL v3. See licenses/GPL-3.txt

#include "ChatForm.h"
#include "FormController.h"
#include "Lang.h"

ChatForm::ChatForm(FormController *controller) : FaceForm(controller), m_message(0), m_hasOptions(false), m_doSetup(true), m_close(false)
{
	Gui::VBox *outer = new Gui::VBox();
	outer->SetSpacing(10.0f);
	Add(outer, 0, 0);

	m_topRegion = new Gui::VBox();
	m_topRegion->SetSpacing(5.0f);
	outer->PackEnd(m_topRegion);

	m_bottomRegion = new Gui::VBox();
	m_bottomRegion->SetSpacing(5.0f);
	outer->PackEnd(m_bottomRegion);
}

void ChatForm::ShowAll()
{
	if (m_doSetup) {
		m_doSetup = false;
		OnOptionClickedTrampoline(0);
	}
	FaceForm::ShowAll();
}

void ChatForm::SetMessage(std::string msg)
{
	if (msg.length() == 0) {
		if (!m_message) return;
		RemoveChild(m_message);
		delete m_message;
		m_message = 0;
		return;
	}

	if (!m_message) {
		m_message = new Gui::Label(msg);
		AddTopWidget(m_message);
		return;
	}

	m_message->SetText(msg);
}

void ChatForm::AddOption(std::string text, int val)
{
	if (!m_hasOptions) {
		AddBottomWidget(new Gui::Label(Lang::SUGGESTED_RESPONSES));
		m_hasOptions = true;
	}

	Gui::Box *box = new Gui::HBox();
	box->SetSpacing(5.0f);

	Gui::Button *b = new Gui::SolidButton();
	b->onClick.connect(sigc::bind(sigc::mem_fun(this, &ChatForm::OnOptionClickedTrampoline), val));

	box->PackEnd(b);
	box->PackEnd(new Gui::Label(text));

	AddBottomWidget(box);
}

void ChatForm::Clear()
{
	m_topRegion->DeleteAllChildren();
	m_bottomRegion->DeleteAllChildren();

	m_message = 0;
	m_hasOptions = false;
}

void ChatForm::Close()
{
	m_close = true;
}

void ChatForm::AddTopWidget(Gui::Widget *w)
{
	m_topRegion->PackEnd(w);
	m_topRegion->ShowAll();
}

void ChatForm::AddBottomWidget(Gui::Widget *w)
{
	m_bottomRegion->PackEnd(w);
	m_bottomRegion->ShowAll();
}

void ChatForm::OnOptionClickedTrampoline(int option)
{
	OnOptionClicked(option);
	if (m_close)
		m_formController->CloseForm();
	else
		m_formController->Refresh();
}

File: data/ui/StationView/ChatForm.lua
-- Copyright © 2008-2014 Pioneer Developers. See AUTHORS.txt for details
-- Licensed under the terms of the GPL v3. See licenses/GPL-3.txt

local Engine = import("Engine")
local Lang = import("Lang")

local Face = import("UI.Game.Face")

local SmallLabeledButton = import("ui/SmallLabeledButton")

local EquipmentTableWidgets = import("EquipmentTableWidgets")

local ui = Engine.ui

local l = Lang.GetResource("ui-core")

local ChatForm = {}

ChatForm.meta = {
	__index = ChatForm,
	class = "ChatForm",
}

function ChatForm.New (chatFunc, removeFunc, ref, tabGroup)
	local form = {
		chatFunc = chatFunc,
		removeFunc = removeFunc,
		ref = ref,
        tabGroup = tabGroup,
	}
	setmetatable(form, ChatForm.meta)
	form.chatFunc(form, 0)
	return form
end

function ChatForm:BuildWidget ()
	local box = ui:VBox(10)

	if self.title then
		box:PackEnd(ui:Label(self.title):SetFont("LARGE"))
	end

	if self.message or self.face then
		local hbox = ui:HBox(5)

		if self.message then
			hbox:PackEnd(ui:Expand("HORIZONTAL", ui:MultiLineText(self.message)))
		end

		if self.face then
			hbox:PackEnd(ui:Align("RIGHT", self.face))
		end

		box:PackEnd(hbox)
	end

	if self.options or self.equipWidgetConfig then

		local scrollBox = ui:VBox()

		if self.options then
			local optionBox = ui:VBox()
			for i = 1,#self.options do
				local option = self.options[i]
				local b = SmallLabeledButton.New(option[1])
				optionBox:PackEnd(b)
				b.button.onClick:Connect(function ()
					if (option[2] == -1) then
						self:Close()
					else
						self.chatFunc(self, option[2])
						ui.layer:SetInnerWidget(self:BuildWidget())
					end
				end)
			end
			scrollBox:PackEnd(optionBox)
		end

		if self.equipWidgetConfig then
			local stationTable, shipTable = EquipmentTableWidgets.Pair(self.equipWidgetConfig)

			scrollBox:PackEnd(
				ui:HBox(10):PackEnd({
					ui:VBox():PackEnd({
						ui:Label(l.AVAILABLE_FOR_PURCHASE):SetFont("HEADING_LARGE"),
						stationTable,
					}),
					ui:VBox():PackEnd({
						ui:Label(l.IN_CARGO_HOLD):SetFont("HEADING_LARGE"),
						shipTable,
					})
				})
			)
		end

		box:PackEnd(ui:Scroller(scrollBox))

	end

	local hangupButton = SmallLabeledButton.New(l.HANG_UP)
	hangupButton.button.onClick:Connect(function () self:Close() end)

	return
		ui:ColorBackground(0,0,0,0.5,
			ui:Grid({1,2,1},{1,2,1}):SetCell(1,1,
				ui:Background(
					ui:VBox(10)
						:PackEnd({ box, hangupButton })
				)
			)
		)
end

function ChatForm:SetTitle (title)
	self.title = title
end

function ChatForm:SetFace (character)

	local faceFlags = {
		character.female and "FEMALE" or "MALE",
		character.armor and "ARMOUR",
	}

	self.face = Face.New(ui, faceFlags, character.seed):SetHeightLines(5)
end

function ChatForm:ClearFace()
	self.face = nil
end

function ChatForm:SetMessage (message)
	self.message = message
end

function ChatForm:AddOption (text, option)
	self.options = self.options or {}
	table.insert(self.options, { text, option })
end

function ChatForm:Clear ()
	self.title = nil
	self.message = nil
	self.options = nil
	self.tradeFuncs = nil
end

local tradeFuncKeys = { "canTrade", "getStock", "getPrice", "onClickBuy", "onClickSell", "bought", "sold" }
function ChatForm:AddGoodsTrader (funcs)
	self.equipWidgetConfig = {
		stationColumns = { "icon", "name", "price", "stock" },
		shipColumns = { "icon", "name", "amount" },
	}

	for i = 1,#tradeFuncKeys do
		local key = tradeFuncKeys[i]
		local fn = funcs[key]
		if fn then
			self.equipWidgetConfig[key] = function (e)
				return fn(self.ref, e)
			end
		end
	end
end

function ChatForm:Close ()
	ui:DropLayer()
end

function ChatForm:GotoPolice ()
    self:Close()
	self.tabGroup:SwitchTo("police")
end

function ChatForm:RemoveAdvertOnClose()
	-- removing it now. has the same effect for the player, without requiring
	-- us to somehow hook the entire view being closed (eg switch back to worldview)
	if self.removeFunc then
		self.removeFunc()
	end
end

return ChatForm
